corehq/apps/api/validation.py

```diff
@@ -0,0 +1,87 @@
+from memoized import memoized
+
+from corehq.apps.custom_data_fields.models import CustomDataFieldsDefinition
+from corehq.apps.reports.util import get_allowed_tableau_groups_for_domain
+from corehq.apps.user_importer.importer import SiteCodeToLocationCache
+from corehq.apps.user_importer.validation import (
+    RoleValidator,
+    ProfileValidator,
+    LocationValidator,
+    TableauGroupsValidator,
+    TableauRoleValidator,
+    CustomDataValidator,
+    EmailValidator,
+)
+from corehq.apps.users.validation import validate_primary_location_assignment
+from corehq.apps.registration.validation import AdminInvitesUserFormValidator
+
+
+class WebUserResourceValidator():
+    def __init__(self, domain, requesting_user):
+        self.domain = domain
+        self.requesting_user = requesting_user
+
+    @property
+    def roles_by_name(self):
+        from corehq.apps.users.views.utils import get_editable_role_choices
+        return {role[1]: role[0] for role in get_editable_role_choices(self.domain, self.requesting_user,
+                                                  allow_admin_role=True)}
+
+    @property
+    @memoized
+    def profiles_by_name(self):
+        from corehq.apps.users.views.mobile.custom_data_fields import UserFieldsView
+        return CustomDataFieldsDefinition.get_profiles_by_name(self.domain, UserFieldsView.field_type)
+
+    @property
+    def location_cache(self):
+        return SiteCodeToLocationCache(self.domain)
+
+    def validate_parameters(self, parameters):
+        allowed_params = ['email', 'role', 'primary_location', 'assigned_locations',
+                          'profile', 'custom_user_data', 'tableau_role', 'tableau_groups']
+        invalid_params = [param for param in parameters if param not in allowed_params]
+        if invalid_params:
+            return f"Invalid parameter(s): {', '.join(invalid_params)}"
+        return AdminInvitesUserFormValidator.validate_parameters(self.domain, self.requesting_user, parameters)
+
+    def validate_role(self, role):
+        spec = {'role': role}
+        return RoleValidator(self.domain, self.roles_by_name()).validate_spec(spec)
+
+    def validate_profile(self, new_profile_name):
+        profile_validator = ProfileValidator(self.domain, self.requesting_user, True, self.profiles_by_name())
+        spec = {'user_profile': new_profile_name}
+        return profile_validator.validate_spec(spec)
+
+    def validate_custom_data(self, custom_data, profile_name):
+        custom_data_validator = CustomDataValidator(self.domain, self.profiles_by_name())
+        spec = {'data': custom_data, 'user_profile': profile_name}
+        return custom_data_validator.validate_spec(spec)
+
+    def validate_email(self, email, is_post):
+        if is_post:
```

Comment:
I was confused by that too, but bulk import works a bit differently. Instead of doing validation upfront, it first tries to find the existing user. A new user is only created if no match is found ([see here](https://github.com/dimagi/commcare-hq/blob/6366ebf3ef386ef7ecd15d066ecaac738d171496/corehq/apps/user_importer/importer.py#L728)). Similarly, when [processing an invitation](https://github.com/dimagi/commcare-hq/blob/6366ebf3ef386ef7ecd15d066ecaac738d171496/corehq/apps/user_importer/importer.py#L843), it [checks for an existing invitation](https://github.com/dimagi/commcare-hq/blob/6366ebf3ef386ef7ecd15d066ecaac738d171496/corehq/apps/user_importer/importer.py#L333) and only creates a new one if it doesn't already exist.

So, the bulk import approach is to try and update existing users or invitations, creating new ones only if they don't exist, rather than validating upfront. This wouldn't work in the case of the API (when POSTing) or the invitation form, though, because those expect to create a new invitation upfront. It wouldn't make sense to fall back to updating an existing invitation if one already exists.